✨ feat: support to customize Embedding model with env

cookieY opened the pull request:
#### 💻 变更类型 | Change Type

<!-- For change type, change [ ] to [x]. -->

- [x] ✨ feat
- [ ] 🐛 fix
- [ ] ♻️ refactor
- [ ] 💄 style
- [ ] 👷 build
- [ ] ⚡️ perf
- [ ] 📝 docs
- [ ] 🔨 chore

#### 🔀 变更说明 | Description of Change

- close https://github.com/lobehub/lobe-chat/issues/3785


#### 📝 补充信息 | Additional Information

由于之前代码commit差距过大 重新提交，提交说明参考 #4370 
<!-- Add any other context about the Pull Request here. -->

cookieY committed:
feat: 添加嵌入模型配置支持，更新相关文档和测试

cookieY commented:
> 支不支持本地部署的embedding模型？例如 api 地址是 http://127.0.0.1/v1 的类 openai 接口？

基于现有的框架设计无法实现这样的自定义需求，本地部署可以使用 ollama

arvinxx committed:
Merge branch 'main' into embedding

arvinxx commented on src/config/knowledge.ts:
```diff
@@ -4,10 +4,14 @@ import { z } from 'zod';
     runtimeEnv: {
+      DEFAULT_FILES_CONFIG: !!process.env.DEFAULT_FILES_CONFIG
+        ? process.env.DEFAULT_FILES_CONFIG
+        : 'embedding_model=openai/embedding-text-3-small,reranker_model=cohere/rerank-english-v3.0,query_mode=full_text',
```
这里不应该给默认值

arvinxx commented on src/const/settings/knowledge.ts:
默认的 embedding model 、 query model 以及reranker model 不应该用 DEFAULT_SYSTEM_AGENT_ITEM 吧？

arvinxx commented on src/database/schemas/ragEvals.ts:
这里不能用任何服务端方法。因为 schema 既会在 server 端跑，也会在 client端执行

如果要改默认向量化模型，这里就不需要加 `$defaultFn(() => DEFAULT_EMBEDDING_MODEL),` 了

arvinxx commented on src/libs/agent-runtime/types/embeddings.ts:
```diff
@@ -4,6 +4,7 @@ export interface EmbeddingsPayload {
    * supported in `text-embedding-3` and later models.
    */
   dimensions?: number;
+  index?: number;
```
为啥这里还需要index？

arvinxx commented on src/libs/agent-runtime/utils/openaiCompatibleFactory/index.ts:
```diff
@@ -275,22 +275,46 @@ export const LobeOpenAICompatibleFactory = <T extends Record<string, any> = any>
     }
 
+    /**
+     * Due to the limitation of a maximum of 8k tokens per request in
```
8k 不是 request 的要求，而是 openai embedding 模型的上下文大小，因此这里不应该改变。要改也是上层调用 embedding 的地方做 8k 的分块

arvinxx commented on src/server/globalConfig/parseSystemAgent.ts:
这里的实现逻辑并不合理。system agent 和 embedding 是两个不同的业务功能模块，不应该混在一起。单独写一个 parseFileConfig

arvinxx commented on src/server/modules/Files/index.ts:
这个模块感觉没有太大必要，而且未来提供 ui 配置还没法合并，不应该这么写。

cookieY committed:
feat: 重构文件配置，更新默认设置和相关测试

cookieY committed:
Merge branch 'main' of https://github.com/lobehub/lobe-chat into embedding

cookieY committed:
Merge branch 'embedding' of github.com:cookieY/lobe-chat into embedding

# Conflicts:
#	.env.example

cookieY commented:
@arvinxx 已根据建议全部进行修改

jiongjiongJOJO commented:
期待尽快合并！！

huanyingtianhe commented:
looking forward to seeing the change soon

bestZwei commented:
期待尽快合并！！

arvinxx commented on src/server/routers/async/file.ts:
```diff
@@ -57,6 +55,9 @@ export const fileRouter = router({
 
       const asyncTask = await ctx.asyncTaskModel.findById(input.taskId);
 
+      const model = getServerDefaultFilesConfig().embedding_model.model;
```
```
const{ model , provider }=getServerDefaultFilesConfig().embeddingModel ||  DEFAULT_EMBEDDING_MODEL_ITEM ;
```

arvinxx commented on src/server/routers/lambda/chunk.ts:
同上

arvinxx commented on src/server/routers/async/file.ts:
```diff
@@ -84,25 +85,23 @@ export const fileRouter = router({
                   dimensions: 1024,
                   input: chunks.map((c) => c.text),
-                  model: input.model,
+                  model: model,
```
单个 model 即可

cookieY replied:
这里没太明白什么意思 - -

arvinxx replied:
```
input: chunks.map((c) => c.text),
model,
```

arvinxx commented on src/server/routers/async/file.ts:
```diff
@@ -111,7 +110,7 @@ export const fileRouter = router({
                     embeddings: e,
                     fileId: input.fileId,
-                    model: input.model,
+                    model: model,
```
同上

arvinxx commented on src/server/routers/lambda/chunk.ts:
```diff
@@ -130,27 +127,27 @@ export const chunkRouter = router({
       let embedding: number[];
       let ragQueryId: string;
-
+      console.log('embeddingProvider:', provider);
```
console.log 去掉

arvinxx reviewed (changes requested):
再改一轮，就差不多了

arvinxx committed:
Merge branch 'main' into embedding

cookieY commented on src/const/settings/knowledge.ts:
```diff
@@ -0,0 +1,25 @@
+};
+
+export const DEFAULT_FILES_CONFIG: FilesConfig = {
+  embedding_model: DEFAULT_FILE_EMBEDDING_MODEL_ITEM,
```
在这里我是让配置文件同步也使用小驼峰 还是将配置文件中下划线转写成小驼峰？

arvinxx replied:
配置文件用蛇形，代码里的变量用小驼峰，这样能区分出来

cookieY committed:
♻ Code Refactoring
- Update the file configuration and standardize the model naming to camel case.

cookieY committed:
Merge branch 'embedding' of github.com:cookieY/lobe-chat into embedding

cookieY commented:
@arvinxx 已修改完毕

m0eak commented:
期待尽快合并！！

lwabish commented:
Looking forward to this PR to try involving zhipu-ai embedding-3 to my lobechat

cmx4869 commented:
love you gugs, hope to use custom embedding model soon!!!

zc16607 commented:
能同时支持自定义baseurl和嵌入模型名称吗？

coolangcn commented:
期待

arvinxx committed:
Merge branch 'main' into embedding

arvinxx committed:
Merge branch 'main' into embedding

arvinxx committed:
Merge branch 'main' into embedding

disarticulate commented:
was able to test this out.

After figuring out a correct embedding model for ollama: xiaoming/bce-embedding-base_v1 and trying both the proxy OpenAI Url and the ollama url:
```
DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1
DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1
```
I get the following vague errors:
```
#DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1

embeddingChunks error {
2025-01-19 07:47:39   message: '{"error":{"message":"[object Object]","name":"ResponseError","status_code":500},"errorType":"OllamaBizError","provider":"ollama"}',
2025-01-19 07:47:39   name: 'EmbeddingError'
2025-01-19 07:47:39 }

#DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1
embeddingChunks error {
2025-01-19 07:52:24   message: '{"endpoint":"http://vineyard_ollama:****/v1","error":{"headers":{"content-length":"0","content-type":"application/json; charset=utf-8","date":"Sun, 19 Jan 2025 13:52:24 GMT"},"stack":"Error: 500 status code (no body)\\n    at eV.generate (/app/.next/server/chunks/67863.js:41:23151)\\n    at r3.makeStatusError (/app/.next/server/chunks/67863.js:41:36093)\\n    at r3.makeRequest (/app/.next/server/chunks/67863.js:41:37037)\\n    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)\\n    at async Object.embeddings (/app/.next/server/chunks/9992.js:1:7027)\\n    at async concurrency (/app/.next/server/app/(backend)/trpc/async/[trpc]/route.js:1:4431)\\n    at async /app/.next/server/chunks/42368.js:51:290507","status":500},"errorType":"ProviderBizError","provider":"openai"}',
2025-01-19 07:52:24   name: 'EmbeddingError'
2025-01-19 07:52:24 }

```

It does appear to be chunking, and takes ~a minute to respond. I'm running everything inside a container and have had to reroute some hosts requests.

SPLT12 commented:
> was able to test this out.
> 
> After figuring out a correct embedding model for ollama: xiaoming/bce-embedding-base_v1 and trying both the proxy OpenAI Url and the ollama url:
> 
> ```
> DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1
> DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1
> ```
> 
> I get the following vague errors:
> 
> ```
> #DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1
> 
> embeddingChunks error {
> 2025-01-19 07:47:39   message: '{"error":{"message":"[object Object]","name":"ResponseError","status_code":500},"errorType":"OllamaBizError","provider":"ollama"}',
> 2025-01-19 07:47:39   name: 'EmbeddingError'
> 2025-01-19 07:47:39 }
> 
> #DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1
> embeddingChunks error {
> 2025-01-19 07:52:24   message: '{"endpoint":"http://vineyard_ollama:****/v1","error":{"headers":{"content-length":"0","content-type":"application/json; charset=utf-8","date":"Sun, 19 Jan 2025 13:52:24 GMT"},"stack":"Error: 500 status code (no body)\\n    at eV.generate (/app/.next/server/chunks/67863.js:41:23151)\\n    at r3.makeStatusError (/app/.next/server/chunks/67863.js:41:36093)\\n    at r3.makeRequest (/app/.next/server/chunks/67863.js:41:37037)\\n    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)\\n    at async Object.embeddings (/app/.next/server/chunks/9992.js:1:7027)\\n    at async concurrency (/app/.next/server/app/(backend)/trpc/async/[trpc]/route.js:1:4431)\\n    at async /app/.next/server/chunks/42368.js:51:290507","status":500},"errorType":"ProviderBizError","provider":"openai"}',
> 2025-01-19 07:52:24   name: 'EmbeddingError'
> 2025-01-19 07:52:24 }
> ```
> 
> It does appear to be chunking, and takes ~a minute to respond. I'm running everything inside a container and have had to reroute some hosts requests.

@disarticulate Same here, trying to figure it out but I get always those type of errors.

cookieY commented:
> > was able to test this out.
> > After figuring out a correct embedding model for ollama: xiaoming/bce-embedding-base_v1 and trying both the proxy OpenAI Url and the ollama url:
> > ```
> > DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1
> > DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I get the following vague errors:
> > ```
> > #DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1
> > 
> > embeddingChunks error {
> > 2025-01-19 07:47:39   message: '{"error":{"message":"[object Object]","name":"ResponseError","status_code":500},"errorType":"OllamaBizError","provider":"ollama"}',
> > 2025-01-19 07:47:39   name: 'EmbeddingError'
> > 2025-01-19 07:47:39 }
> > 
> > #DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1
> > embeddingChunks error {
> > 2025-01-19 07:52:24   message: '{"endpoint":"http://vineyard_ollama:****/v1","error":{"headers":{"content-length":"0","content-type":"application/json; charset=utf-8","date":"Sun, 19 Jan 2025 13:52:24 GMT"},"stack":"Error: 500 status code (no body)\\n    at eV.generate (/app/.next/server/chunks/67863.js:41:23151)\\n    at r3.makeStatusError (/app/.next/server/chunks/67863.js:41:36093)\\n    at r3.makeRequest (/app/.next/server/chunks/67863.js:41:37037)\\n    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)\\n    at async Object.embeddings (/app/.next/server/chunks/9992.js:1:7027)\\n    at async concurrency (/app/.next/server/app/(backend)/trpc/async/[trpc]/route.js:1:4431)\\n    at async /app/.next/server/chunks/42368.js:51:290507","status":500},"errorType":"ProviderBizError","provider":"openai"}',
> > 2025-01-19 07:52:24   name: 'EmbeddingError'
> > 2025-01-19 07:52:24 }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It does appear to be chunking, and takes ~a minute to respond. I'm running everything inside a container and have had to reroute some hosts requests.
> 
> @disarticulate Same here, trying to figure it out but I get always those type of errors.

The issue seems to be related to not considering multiple slashes (/) when parsing configuration data. I will submit a PR to resolve this issue.

disarticulate commented:
I also discovered a ollama error that returns errots based on embedding
ctx. I got it to work wiyh ollama/quentinz/bge-large-zh-v1.5:latest

On Mon, Jan 20, 2025, 06:47 Henry Yee ***@***.***> wrote:

> was able to test this out.
> After figuring out a correct embedding model for ollama:
> xiaoming/bce-embedding-base_v1 and trying both the proxy OpenAI Url and the
> ollama url:
>
> DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1
> DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1
>
> I get the following vague errors:
>
> #DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1
>
> embeddingChunks error {
> 2025-01-19 07:47:39   message: '{"error":{"message":"[object Object]","name":"ResponseError","status_code":500},"errorType":"OllamaBizError","provider":"ollama"}',
> 2025-01-19 07:47:39   name: 'EmbeddingError'
> 2025-01-19 07:47:39 }
>
> #DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1
> embeddingChunks error {
> 2025-01-19 07:52:24   message: '{"endpoint":"http://vineyard_ollama:****/v1","error":{"headers":{"content-length":"0","content-type":"application/json; charset=utf-8","date":"Sun, 19 Jan 2025 13:52:24 GMT"},"stack":"Error: 500 status code (no body)\\n    at eV.generate (/app/.next/server/chunks/67863.js:41:23151)\\n    at r3.makeStatusError (/app/.next/server/chunks/67863.js:41:36093)\\n    at r3.makeRequest (/app/.next/server/chunks/67863.js:41:37037)\\n    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)\\n    at async Object.embeddings (/app/.next/server/chunks/9992.js:1:7027)\\n    at async concurrency (/app/.next/server/app/(backend)/trpc/async/[trpc]/route.js:1:4431)\\n    at async /app/.next/server/chunks/42368.js:51:290507","status":500},"errorType":"ProviderBizError","provider":"openai"}',
> 2025-01-19 07:52:24   name: 'EmbeddingError'
> 2025-01-19 07:52:24 }
>
> It does appear to be chunking, and takes ~a minute to respond. I'm running
> everything inside a container and have had to reroute some hosts requests.
>
> @disarticulate <https://github.com/disarticulate> Same here, trying to
> figure it out but I get always those type of errors.
>
> The issue seems to be related to not considering multiple slashes (/) when
> parsing configuration data. I will submit a PR to resolve this issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lobehub/lobe-chat/pull/5177#issuecomment-2602339840>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEFHWPPWTM2QSHRSZMCIRAL2LTV6FAVCNFSM6AAAAABUF2LTLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBSGMZTSOBUGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

SPLT12 commented:
Mine is showing this
![image](https://github.com/user-attachments/assets/3d6b11c7-80fd-440b-9e31-262f0bbf56a6)


> I also discovered a ollama error that returns errots based on embedding ctx. I got it to work wiyh ollama/quentinz/bge-large-zh-v1.5:latest
> […](#)
> On Mon, Jan 20, 2025, 06:47 Henry Yee ***@***.***> wrote: was able to test this out. After figuring out a correct embedding model for ollama: xiaoming/bce-embedding-base_v1 and trying both the proxy OpenAI Url and the ollama url: DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1 DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1 I get the following vague errors: #DEFAULT_FILES_CONFIG=embedding_model=ollama/xiaoming/bce-embedding-base_v1 embeddingChunks error { 2025-01-19 07:47:39 message: '{"error":{"message":"[object Object]","name":"ResponseError","status_code":500},"errorType":"OllamaBizError","provider":"ollama"}', 2025-01-19 07:47:39 name: 'EmbeddingError' 2025-01-19 07:47:39 } #DEFAULT_FILES_CONFIG=embedding_model=openai/xiaoming/bce-embedding-base_v1 embeddingChunks error { 2025-01-19 07:52:24 message: '{"endpoint":"http://vineyard_ollama:****/v1","error":{"headers":{"content-length":"0","content-type":"application/json; charset=utf-8","date":"Sun, 19 Jan 2025 13:52:24 GMT"},"stack":"Error: 500 status code (no body)\\n at eV.generate (/app/.next/server/chunks/67863.js:41:23151)\\n at r3.makeStatusError (/app/.next/server/chunks/67863.js:41:36093)\\n at r3.makeRequest (/app/.next/server/chunks/67863.js:41:37037)\\n at process.processTicksAndRejections (node:internal/process/task_queues:105:5)\\n at async Object.embeddings (/app/.next/server/chunks/9992.js:1:7027)\\n at async concurrency (/app/.next/server/app/(backend)/trpc/async/[trpc]/route.js:1:4431)\\n at async /app/.next/server/chunks/42368.js:51:290507","status":500},"errorType":"ProviderBizError","provider":"openai"}', 2025-01-19 07:52:24 name: 'EmbeddingError' 2025-01-19 07:52:24 } It does appear to be chunking, and takes ~a minute to respond. I'm running everything inside a container and have had to reroute some hosts requests. @disarticulate <https://github.com/disarticulate> Same here, trying to figure it out but I get always those type of errors. The issue seems to be related to not considering multiple slashes (/) when parsing configuration data. I will submit a PR to resolve this issue. — Reply to this email directly, view it on GitHub <[#5177 (comment)](https://github.com/lobehub/lobe-chat/pull/5177#issuecomment-2602339840)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEFHWPPWTM2QSHRSZMCIRAL2LTV6FAVCNFSM6AAAAABUF2LTLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBSGMZTSOBUGA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

steinvenic commented:
![image](https://github.com/user-attachments/assets/022e984e-1a75-4ebe-a049-edcdfe6af298)
![image](https://github.com/user-attachments/assets/0f7c3dc2-bc44-49d0-98f7-6b2ee47b5823)
```
DEFAULT_FILES_CONFIG="embedding_model=ollama/xiaoming/bce-embedding-base_v1"
```
